site/source/docs/optimizing/Optimizing-Code.rst

```diff
@@ -90,6 +89,20 @@ The following compiler settings can help (see ``src/settings.js`` for more detai
 - The ``ENVIRONMENT`` flag lets you specify that the output will only run on the web, or only run in node.js, etc. This prevents the compiler from emitting code to support all possible runtime environments, saving ~2KB.
 - You can use ``ELIMINATE_DUPLICATE_FUNCTIONS`` to remove duplicate functions, which C++ templates often create. (This is already done by default for wasm, in ``-O1`` and above.)
 
+LTO
+===
+
+Link Time Optimization (LTO) lets the compiler do more work, as it can inline across separate compilation units, and even with system libraries. The :ref:`main relevant flag <emcc-llvm-lto>` is ``--llvm-lto 1`` at link time.
+
+Separately from that flag, the linker must also receive LLVM bitcode files in order to run LTO on them. With fastcomp that is always the case; with the LLVM wasm backend, object files main contain either wasm or bitcode. The linker can handle a mix of the two, but can only do LTO on the bitcode files. You can control that with the following flags:
+
+- The ``-flto`` flag tells the compiler to emit bitcode in object files, but does *not* affect system libraries.
+- The ``-s WASM_OBJECT_FILES=0`` also flag tells the compiler to emit bitcode to object files, and also to use bitcode system libraries.
+
+Thus, to allow maximal LTO opportunities with the LLVM wasm backend, build all source files with ``-s WASM_OBJECT_FILES=0`` and link with ``-s WASM_OBJECT_FILES=0 --llvm-lto 1``.
```

Comment:
This is OK as a documentation of the existing status quo, but it's still pretty awkward and there are a lot of combinations of flags/behavior, not all of which are not useful. So it would be nice to simplify or line up with native a bit better. For example WASM_OBJECT_FILES is a legacy and it controls a lot of code that we can rip out once fastcomp is gone (we should be able to do what native platforms do and use lld to drive the LTO pipeline rather than driver code). Also if the only delta from -flto to NATIVE_OBJECT_FILES is that it also affects system libraries, then the name should reflect that, and it should only be required/allowed at link time. And --llvm-lto is a meaningless name when everything is done with LLVM.

Reply:
Agreed about simplifying this stuff later, yes.